From: Niko Tyni <ntyni@debian.org>
Subject: fix a segmentation fault with array ties
Closes: 525180

[perl #51636]

The bug triggers with Gtk2::SimpleList, as discussed in
 http://www.xray.mpe.mpg.de/mailing-lists/perl5-porters/2009-04/msg00368.html

Upstream changes 
 33495 b63c7c552a2e9cf2b2c5eb492358b8567fd16179
 33778 90630e3c741716305d7f1da4df5eab5c1bee42cc

---
 av.c     |    2 +-
 pp_hot.c |   22 +++++++++++++++++++---
 2 files changed, 20 insertions(+), 4 deletions(-)

diff --git a/av.c b/av.c
index 0d46117..459050a 100644
--- a/av.c
+++ b/av.c
@@ -433,7 +433,7 @@ Perl_av_clear(pTHX_ register AV *av)
     /* Give any tie a chance to cleanup first */
     if (SvRMAGICAL(av)) {
 	const MAGIC* const mg = SvMAGIC(av);
-	if (PL_delaymagic && mg->mg_type == PERL_MAGIC_isa)
+	if (PL_delaymagic && mg && mg->mg_type == PERL_MAGIC_isa)
 	    PL_delaymagic |= DM_ARRAY;
         else
 	    mg_clear((SV*)av); 
diff --git a/pp_hot.c b/pp_hot.c
index 57fa328..0957d3c 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -1028,8 +1028,14 @@ PP(pp_aassign)
 		*(relem++) = sv;
 		didstore = av_store(ary,i++,sv);
 		if (magic) {
-		    if (SvSMAGICAL(sv))
+		    if (SvSMAGICAL(sv)) {
+			/* More magic can happen in the mg_set callback, so we
+			 * backup the delaymagic for now. */
+			U16 dmbak = PL_delaymagic;
+			PL_delaymagic = 0;
 			mg_set(sv);
+			PL_delaymagic = dmbak;
+		    }
 		    if (!didstore)
 			sv_2mortal(sv);
 		}
@@ -1059,8 +1065,12 @@ PP(pp_aassign)
 			duplicates += 2;
 		    didstore = hv_store_ent(hash,sv,tmpstr,0);
 		    if (magic) {
-			if (SvSMAGICAL(tmpstr))
+			if (SvSMAGICAL(tmpstr)) {
+			    U16 dmbak = PL_delaymagic;
+			    PL_delaymagic = 0;
 			    mg_set(tmpstr);
+			    PL_delaymagic = dmbak;
+			}
 			if (!didstore)
 			    sv_2mortal(tmpstr);
 		    }
@@ -1084,7 +1094,13 @@ PP(pp_aassign)
 	    }
 	    else
 		sv_setsv(sv, &PL_sv_undef);
-	    SvSETMAGIC(sv);
+
+	    if (SvSMAGICAL(sv)) {
+		U16 dmbak = PL_delaymagic;
+		PL_delaymagic = 0;
+		mg_set(sv);
+		PL_delaymagic = dmbak;
+	    }
 	    break;
 	}
     }
-- 
tg: (71b2123..) fixes/tied-array-crash (depends on: upstream)
